compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnnotationReference.kt

```diff
@@ -41,12 +51,46 @@ public sealed class AnnotationReference {
   public class Psi internal constructor(
     public val annotation: KtAnnotationEntry,
     override val classReference: ClassReference
-  ) : AnnotationReference()
+  ) : AnnotationReference() {
+
+    override fun isQualifier(): Boolean {
+      return annotation
+        .typeReference
+        ?.requireFqName(module)
+        // Often entries are annotated with @Inject, in this case we know it's not a qualifier and
+        // we can stop early.
+        ?.takeIf { it != injectFqName }
+        ?.classDescriptor(module)
+        ?.annotations
+        ?.hasAnnotation(qualifierFqName)
+        ?: false
+    }
+
+    override fun isMapKey(): Boolean {
+      return annotation
+        .typeReference
+        ?.requireFqName(module)
+        ?.takeIf { it != injectFqName }
+        ?.classDescriptor(module)
+        ?.annotations
+        ?.hasAnnotation(mapKeyFqName)
+        ?: false
+    }
+  }
 
   public class Descriptor internal constructor(
     public val annotation: AnnotationDescriptor,
     override val classReference: ClassReference
-  ) : AnnotationReference()
+  ) : AnnotationReference() {
+
+    override fun isQualifier(): Boolean {
+      return annotation.annotationClass?.annotations?.hasAnnotation(qualifierFqName) ?: false
+    }
+
+    override fun isMapKey(): Boolean {
+      return annotation.annotationClass?.annotations?.hasAnnotation(MapKey::class.fqName) ?: false
```

Comment:
Same here. Also use `mapKeyFqName` instead of `MapKey::class.fqName` to avoid creating unnecessary instances.